src/lib.rs

```diff
@@ -1190,20 +1209,80 @@ pub enum MapMode {
     Write,
 }
 
+fn range_to_offset_size<S: RangeBounds<BufferAddress>>(bounds: S) -> (BufferAddress, BufferSize) {
+    let offset = match bounds.start_bound() {
+        Bound::Included(&bound) => bound,
+        Bound::Excluded(&bound) => bound + 1,
+        Bound::Unbounded => 0,
+    };
+    let size = match bounds.end_bound() {
+        Bound::Included(&bound) => BufferSize(bound + 1 - offset),
+        Bound::Excluded(&bound) => BufferSize(bound - offset),
+        Bound::Unbounded => BufferSize::WHOLE,
+    };
+
+    (offset, size)
+}
+
+pub struct BufferView<'a> {
+    buffer: &'a Buffer,
+    data: &'a [u8],
+    offset: BufferAddress,
+    size: BufferSize,
+}
+
+pub struct BufferViewMut<'a> {
+    buffer: &'a Buffer,
+    data: &'a mut [u8],
+    offset: BufferAddress,
+    size: BufferSize,
+}
+
+impl std::ops::Deref for BufferView<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        self.data
+    }
+}
+
+impl std::ops::Deref for BufferViewMut<'_> {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        self.data
+    }
+}
+
+impl std::ops::DerefMut for BufferViewMut<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.data
+    }
+}
+
+impl Drop for BufferView<'_> {
+    fn drop(&mut self) {
+        self.buffer
```

Comment:
let's put this inside `if !thread::panicking() {}` check

Reply:
Okay, but what does this do?

Reply:
I suppose it's not actually needed here, so nevermind :)

To clarify, if a drop() method does something non-trivial, like calling into `wgpu`, we don't want that to happen if the drop is calling during stack unwinding of another panic. That would be double panic, and it's annoying to look at :)